[CIR] Upstream minimal builtin function call support

mmha opened the pull request:
This patch adds all bits required to implement builtin function calls to ClangIR. It doesn't actually implement any of the builtins except those that fold to a constant ahead of CodeGen (`__builtin_is_constant_evaluated()` being one example). A few builtins that receive special handling like FMA are still kept in this patch and explicitly marked as NYI so we don't have to backtrack those later on. This increased the patch size unfortunately.

It also adds the `LLVMIntrinsicCallOp` instruction to the CIR dialect.

mmha committed:
clang-format

mmha committed:
more clang-format, remove duplicated code

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -0,0 +1,453 @@
+    }
+    if (result.Val.isFloat()) {
+      // Note: we are using result type of CallExpr to determine the type of
+      // the constant. Clang Codegen uses the result value to make judgement
```
```suggestion
      // the constant. Classic codegen uses the result value to determine
```
and delete 'of' at the start of the next line.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -201,6 +201,19 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
+
+  cir::ConstantOp getConstFP(mlir::Location loc, mlir::Type t,
+                             llvm::APFloat fpVal) {
+    assert((mlir::isa<cir::SingleType, cir::DoubleType>(t)) &&
```
This doesn't seem like a good assumption, and it's crashing the incubator in this case: https://godbolt.org/z/Ycv1Gv7Y1.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -0,0 +1,453 @@
+      // the constant. Clang Codegen uses the result value to make judgement
+      // of the type. We feel it should be Ok to use expression type because
+      // it is hard to imagine a builtin function evaluates to
+      // a value that over/underflows its own defined type.
```
The equivalent code in the classic codegen now looks like this:
```
    if (Result.Val.isFloat())
      return RValue::get(llvm::ConstantFP::get(getLLVMContext(),
                                               Result.Val.getFloat()));
```
Can we make a version of `getConstFP` that takes the APFloat result and deduces the type from that? The result should be the same. I think the type of `result` will always be the same as the type of the expression, but it would be nice to have consistency between the int and float handling here.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -0,0 +1,453 @@
+  // after backend supports IEEE 128-bit style libcalls.
+  if (getTarget().getTriple().isPPC64() &&
+      &getTarget().getLongDoubleFormat() == &llvm::APFloat::IEEEquad()) {
+    cgm.errorNYI("long double builtin mutation");
```
Should this be (not) happening before the constant folding above?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -0,0 +1,453 @@
+          constWithoutErrnoOrExceptions && errnoOverridenToFalseWithOpt;
+  }
+
+  if (generateIntrinsics) {
```
The code to compute the `generateIntrinsics` condition (from line 167 to here) was extremely tricky to get right, and I wouldn't be surprised if it changes as new pragmas are added. This really should be shared with classic codegen. Can you add a comment to that effect?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -0,0 +1,453 @@
+    return {};
+  }
+
+  switch (builtinIDIfNoAsmLabel) {
```
Is there an errorNYI missing here?

erichkeane reviewed (commented):
There seems to be a LOT of code here that isn't in the tests?  The only thing tested is no argument types, and just an int return type, yet there is a ton of builtin code to do a lot more than that. 

I think this needs a lot more work to make sure we're properly testing this.

andykaylor reviewed (commented):
This PR needs to either remove a lot of code or add a lot more test cases.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -201,6 +201,19 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
 
+  cir::ConstantOp getConstInt(mlir::Location loc, llvm::APSInt intVal);
+
+  cir::ConstantOp getConstInt(mlir::Location loc, llvm::APInt intVal);
```
We need to handle APSInt to get signed constants represented correctly.

erichkeane replied:
Actually, we probably only have to handle `APSInt`.  It has an (explicit) ctor from `APInt`.  Wonder if we should just implement one in terms of the other.

mmha committed:
[CIR] Upstream minimal builtin function call support

This patch adds all bits required to implement builtin function calls to
ClangIR. It doesn't actually implement any of the builtins except those
that fold to a constant ahead of CodeGen
(__builtin_is_constant_evaluated() being one example).

It also adds the LLVMIntrinsicCallOp instruction to the CIR dialect.

mmha committed:
- Remove all builtin handling not needed for constant foldable builtin calls
- Address review feedback

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenBuilder.cpp:
```diff
@@ -38,3 +38,25 @@ mlir::Value CIRGenBuilderTy::getArrayElement(mlir::Location arrayLocBegin,
+
+cir::ConstantOp CIRGenBuilderTy::getConstInt(mlir::Location loc,
+                                             llvm::APInt intVal) {
+  unsigned width = intVal.getBitWidth();
```
```suggestion
  return getConstInt(loc, llvm::APSInt(intVal));
```

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -229,6 +230,36 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
+
+  cir::ConstantOp getConstInt(mlir::Location loc, mlir::Type t, uint64_t c);
+
+  mlir::Type getFPType(llvm::fltSemantics const &sem) {
```
Where is this being used in this patch?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -229,6 +230,36 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
+    }
+  }
+
+  cir::ConstantOp getConstFP(mlir::Location loc, llvm::APFloat fpVal) {
```
This too?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -0,0 +1,55 @@
+      return RValue::get(builder.getConstInt(getLoc(e->getSourceRange()),
+                                             result.Val.getInt()));
+    }
+    if (result.Val.isFloat()) {
```
Do we have a test for one that has a floating point result?

erichkeane commented on clang/test/CIR/CodeGen/builtin_call.cpp:
```diff
@@ -0,0 +1,30 @@
+// RUN: %clang_cc1 -std=c++11 -triple x86_64-unknown-linux-gnu -Wno-unused-value -emit-llvm %s -o %t.ll
+// RUN: FileCheck --input-file=%t.ll %s -check-prefix=OGCG
+
+constexpr extern int cx_var = __builtin_is_constant_evaluated();
```
We still seem to have a lot of code not tested in here?  I don't see any FP stuff here, for example.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -1004,8 +1004,48 @@ static cir::FuncOp emitFunctionDeclPointer(CIRGenModule &cgm, GlobalDecl gd) {
 }
 
-static CIRGenCallee emitDirectCallee(CIRGenModule &cgm, GlobalDecl gd) {
-  assert(!cir::MissingFeatures::opCallBuiltinFunc());
```
I believe you've removed all instances of this MissingFeature. It should be removed from the header.

mmha committed:
Add FP tests

andykaylor reviewed (approved):
lgtm

Lancern commented on clang/test/CIR/CodeGen/builtin_call.cpp:
```diff
@@ -0,0 +1,78 @@
+// OGCG: @cx_var_ld = {{.*}} x86_fp80 0xK7FFF8000000000000000
+
+int is_constant_evaluated() {
+    return __builtin_is_constant_evaluated();
```
nit: reduce indentation

Lancern commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -229,6 +231,19 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
+
+  cir::ConstantOp getConstInt(mlir::Location loc, llvm::APInt intVal);
+
+  cir::ConstantOp getConstInt(mlir::Location loc, mlir::Type t, uint64_t c);
+
+  cir::ConstantOp getConstFP(mlir::Location loc, mlir::Type t,
+                             llvm::APFloat fpVal) {
+    assert(mlir::isa<cir::CIRFPTypeInterface>(t) &&
+           "expected floating point type");
+    return create<cir::ConstantOp>(loc, getAttr<cir::FPAttr>(t, fpVal));
+  }
```
The implementation of `getConstInt` and `getConstFP` looks very similar, yet `getConstInt` is defined in `CIRGenBuilder.cpp` while `getConstFP` is defined in the header. Can we make it consistent?

Lancern reviewed (approved):
LGTM except 2 nits

mmha committed:
Address more review feedback
